File: SPOS/New Codes/FCFS.py
"""
Very simple FCFS (First Come First Serve) scheduler
Non-preemptive, based on arrival time. Menu driven for beginners.
"""

def run_fcfs():
    try:
        n = int(input("Enter number of processes: ").strip())
        if n <= 0:
            print("Number of processes must be positive.")
            return
    except ValueError:
        print("Please enter a valid integer.")
        return

    procs = []
    for i in range(n):
        while True:
            try:
                a = int(input(f"Arrival time for P{i}: ").strip())
                b = int(input(f"Burst time for P{i}: ").strip())
                if a < 0 or b <= 0:
                    print("Arrival must be >= 0 and Burst must be > 0. Try again.")
                    continue
                break
            except ValueError:
                print("Please enter integers for times. Try again.")
        procs.append({"id": i, "arrival": a, "burst": b})

    # sort by arrival time (stable)
    procs.sort(key=lambda p: p["arrival"])

    time = 0
    order = []
    # compute metrics
    for p in procs:
        if time < p["arrival"]:
            order.append(f"IDLE({p['arrival'] - time})")
            time = p["arrival"]
        p_response = time - p["arrival"]
        order.append(f"P{p['id']}")
        time += p["burst"]
        p_completion = time
        p_turnaround = p_completion - p["arrival"]
        p_waiting = p_turnaround - p["burst"]
        # attach metrics to process dict for printing
        p.update({
            "completion": p_completion,
            "turnaround": p_turnaround,
            "waiting": p_waiting,
            "response": p_response
        })

    # print tables and averages
    print("\nINPUT TABLE")
    print(f"{'Proc':<6}{'Arrival':<8}{'Burst':<6}")
    for p in procs:
        print(f"P{p['id']:<5}{p['arrival']:<8}{p['burst']:<6}")

    print("\nEXECUTION ORDER")
    print(" -> ".join(order))

    print("\nRESULTS")
    print(f"{'Proc':<6}{'Comp':<6}{'TAT':<6}{'Wait':<6}{'Resp':<6}")
    total_comp = total_tat = total_wait = total_resp = 0
    for p in procs:
        print(f"P{p['id']:<5}{p['completion']:<6}{p['turnaround']:<6}{p['waiting']:<6}{p['response']:<6}")
        total_comp += p["completion"]
        total_tat += p["turnaround"]
        total_wait += p["waiting"]
        total_resp += p["response"]

    m = len(procs)
    print("\nAVERAGES")
    print(f"Average Completion Time : {total_comp / m}")
    print(f"Average Turnaround Time : {total_tat / m}")
    print(f"Average Waiting Time    : {total_wait / m}")
    print(f"Average Response Time   : {total_resp / m}")

    # simple Gantt-like timeline
    print("\nGANTT TIMELINE")
    times = []
    parts = []
    time = 0
    for p in procs:
        if time < p["arrival"]:
            parts.append("IDLE")
            times.append(time)
            time = p["arrival"]
        parts.append(f"P{p['id']}")
        times.append(time)
        time = p["completion"]
    times.append(time)
    print(" | ".join(parts))
    print("  ".join(str(t) for t in times))


def main_menu():
    while True:
        print("\n===== FCFS Scheduler =====")
        print("1. Run FCFS scheduling")
        print("2. Exit")
        choice = input("Enter choice (1 or 2): ").strip()
        if choice == "1":
            run_fcfs()
        elif choice == "2":
            print("Goodbye!")
            break
        else:
            print("Invalid choice. Enter 1 or 2.")

if __name__ == "__main__":
    main_menu()

File: SPOS Java Codes/FCFS.java
import java.util.*;

public class FCFS {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of processes: ");
        int n = sc.nextInt();

        int[] pid = new int[n]; // Process IDs
        int[] arrival = new int[n]; // Arrival times
        int[] burst = new int[n]; // Burst times
        int[] completion = new int[n]; // Completion times
        int[] turnaround = new int[n]; // Turnaround times
        int[] waiting = new int[n]; // Waiting times

        // Input process details
        for (int i = 0; i < n; i++) {
            System.out.print("Enter arrival time for process " + (i + 1) + ": ");
            arrival[i] = sc.nextInt();
            System.out.print("Enter burst time for process " + (i + 1) + ": ");
            burst[i] = sc.nextInt();
            pid[i] = i + 1;
        }

        // Sort processes by arrival time (FCFS)
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (arrival[j] > arrival[j + 1]) {
                    // Swap arrival times
                    int temp = arrival[j];
                    arrival[j] = arrival[j + 1];
                    arrival[j + 1] = temp;

                    // Swap burst times
                    temp = burst[j];
                    burst[j] = burst[j + 1];
                    burst[j + 1] = temp;

                    // Swap process IDs
                    temp = pid[j];
                    pid[j] = pid[j + 1];
                    pid[j + 1] = temp;
                }
            }
        }

        // Calculate completion times
        int currentTime = 0;
        for (int i = 0; i < n; i++) {
            if (currentTime < arrival[i]) {
                currentTime = arrival[i];
            }
            completion[i] = currentTime + burst[i];
            currentTime = completion[i];
        }

        // Calculate turnaround and waiting times
        float avgWait = 0, avgTurnaround = 0;
        for (int i = 0; i < n; i++) {
            turnaround[i] = completion[i] - arrival[i];
            waiting[i] = turnaround[i] - burst[i];
            avgWait += waiting[i];
            avgTurnaround += turnaround[i];
        }

        // Display results
        System.out.println("\nProcess Table:");
        System.out.println("PID\tArrival\tBurst\tCompletion\tTurnaround\tWaiting");
        for (int i = 0; i < n; i++) {
            System.out.printf("%d\t%d\t%d\t%d\t\t%d\t\t%d\n",
                    pid[i], arrival[i], burst[i], completion[i], turnaround[i], waiting[i]);
        }

        System.out.printf("\nAverage Turnaround Time: %.2f\n", (avgTurnaround / n));
        System.out.printf("Average Waiting Time: %.2f\n", (avgWait / n));

        sc.close();
    }
}